File: ui/src/models/announcement.ts
import { api } from '@/stores/_config'

export type AnnouncementScopeType = 'world' | 'lobby'
export type AnnouncementStatus = 'draft' | 'published' | 'archived'
export type AnnouncementPopupMode = 'none' | 'once_per_version' | 'every_entry'
export type AnnouncementContentFormat = 'plain' | 'rich'
export type AnnouncementReminderScope = 'lobby_only' | 'site_wide'

export interface AnnouncementItem {
  id: string
  createdAt: string
  updatedAt: string
  scopeType: AnnouncementScopeType
  scopeId: string
  title: string
  content: string
  contentFormat: AnnouncementContentFormat
  status: AnnouncementStatus
  isPinned: boolean
  pinOrder: number
  popupMode: AnnouncementPopupMode
  reminderScope: AnnouncementReminderScope
  requireAck: boolean
  version: number
  publishedAt?: string | null
  createdBy: string
  updatedBy: string
  creatorName?: string
  updaterName?: string
  lastSeenVersion: number
  ackVersion: number
  ackCount: number
  isAcked: boolean
  needsAck: boolean
  canEdit: boolean
}

export interface AnnouncementListResponse {
  items: AnnouncementItem[]
  total: number
  page: number
  pageSize: number
}

export interface AnnouncementPayload {
  title: string
  content: string
  contentFormat: AnnouncementContentFormat
  status: AnnouncementStatus
  isPinned: boolean
  pinOrder: number
  popupMode: AnnouncementPopupMode
  reminderScope: AnnouncementReminderScope
  requireAck: boolean
}

export async function fetchWorldAnnouncements(worldId: string, params?: { page?: number; pageSize?: number; includeAll?: boolean; includeArchived?: boolean }) {
  const { data } = await api.get<AnnouncementListResponse>(`/api/v1/worlds/${worldId}/announcements`, { params })
  return data
}

export async function createWorldAnnouncement(worldId: string, payload: AnnouncementPayload) {
  const { data } = await api.post<{ item: AnnouncementItem }>(`/api/v1/worlds/${worldId}/announcements`, payload)
  return data.item
}

export async function updateWorldAnnouncement(worldId: string, announcementId: string, payload: AnnouncementPayload) {
  const { data } = await api.patch<{ item: AnnouncementItem }>(`/api/v1/worlds/${worldId}/announcements/${announcementId}`, payload)
  return data.item
}

export async function deleteWorldAnnouncement(worldId: string, announcementId: string) {
  await api.delete(`/api/v1/worlds/${worldId}/announcements/${announcementId}`)
}

export async function fetchWorldPendingAnnouncement(worldId: string) {
  const { data } = await api.get<{ item: AnnouncementItem | null }>(`/api/v1/worlds/${worldId}/announcements/pending-popup`)
  return data.item
}

export async function markWorldAnnouncementPopup(worldId: string, announcementId: string) {
  const { data } = await api.post<{ item: AnnouncementItem }>(`/api/v1/worlds/${worldId}/announcements/${announcementId}/mark-popup`)
  return data.item
}

export async function ackWorldAnnouncement(worldId: string, announcementId: string) {
  const { data } = await api.post<{ item: AnnouncementItem }>(`/api/v1/worlds/${worldId}/announcements/${announcementId}/ack`)
  return data.item
}

export async function fetchLobbyAnnouncements(params?: { page?: number; pageSize?: number; includeAll?: boolean; includeArchived?: boolean }) {
  const { data } = await api.get<AnnouncementListResponse>('/api/v1/lobby-announcements', { params })
  return data
}

export async function fetchLobbyPendingAnnouncement(params?: { reminderScope?: AnnouncementReminderScope }) {
  const { data } = await api.get<{ item: AnnouncementItem | null }>('/api/v1/lobby-announcements/pending-popup', { params })
  return data.item
}

export async function markLobbyAnnouncementPopup(announcementId: string) {
  const { data } = await api.post<{ item: AnnouncementItem }>(`/api/v1/lobby-announcements/${announcementId}/mark-popup`)
  return data.item
}

export async function createLobbyAnnouncement(payload: AnnouncementPayload) {
  const { data } = await api.post<{ item: AnnouncementItem }>('/api/v1/admin/lobby-announcements', payload)
  return data.item
}

export async function updateLobbyAnnouncement(announcementId: string, payload: AnnouncementPayload) {
  const { data } = await api.patch<{ item: AnnouncementItem }>(`/api/v1/admin/lobby-announcements/${announcementId}`, payload)
  return data.item
}

export async function deleteLobbyAnnouncement(announcementId: string) {
  await api.delete(`/api/v1/admin/lobby-announcements/${announcementId}`)
}

File: model/announcement.go
package model

import (
	"strings"
	"time"
)

type AnnouncementScopeType string

const (
	AnnouncementScopeWorld AnnouncementScopeType = "world"
	AnnouncementScopeLobby AnnouncementScopeType = "lobby"
)

type AnnouncementStatus string

const (
	AnnouncementStatusDraft     AnnouncementStatus = "draft"
	AnnouncementStatusPublished AnnouncementStatus = "published"
	AnnouncementStatusArchived  AnnouncementStatus = "archived"
)

type AnnouncementContentFormat string

const (
	AnnouncementContentPlain AnnouncementContentFormat = "plain"
	AnnouncementContentRich  AnnouncementContentFormat = "rich"
)

type AnnouncementPopupMode string

const (
	AnnouncementPopupNone           AnnouncementPopupMode = "none"
	AnnouncementPopupOncePerVersion AnnouncementPopupMode = "once_per_version"
	AnnouncementPopupEveryEntry     AnnouncementPopupMode = "every_entry"
)

type AnnouncementReminderScope string

const (
	AnnouncementReminderScopeLobbyOnly AnnouncementReminderScope = "lobby_only"
	AnnouncementReminderScopeSiteWide  AnnouncementReminderScope = "site_wide"
)

type AnnouncementModel struct {
	StringPKBaseModel
	ScopeType     AnnouncementScopeType     `json:"scopeType" gorm:"size:16;index:idx_ann_scope,priority:1"`
	ScopeID       string                    `json:"scopeId" gorm:"size:100;index:idx_ann_scope,priority:2"`
	Title         string                    `json:"title" gorm:"size:120;not null"`
	Content       string                    `json:"content" gorm:"type:text"`
	ContentFormat AnnouncementContentFormat `json:"contentFormat" gorm:"size:16;default:'rich'"`
	Status        AnnouncementStatus        `json:"status" gorm:"size:16;default:'draft';index"`
	IsPinned      bool                      `json:"isPinned" gorm:"default:false;index"`
	PinOrder      int                       `json:"pinOrder" gorm:"default:0;index"`
	ShowInTicker  bool                      `json:"showInTicker" gorm:"default:false;index"`
	PopupMode     AnnouncementPopupMode     `json:"popupMode" gorm:"size:24;default:'none'"`
	ReminderScope AnnouncementReminderScope `json:"reminderScope" gorm:"size:24;default:'lobby_only'"`
	RequireAck    bool                      `json:"requireAck" gorm:"default:false"`
	Version       int                       `json:"version" gorm:"default:1"`
	PublishedAt   *time.Time                `json:"publishedAt"`
	CreatedBy     string                    `json:"createdBy" gorm:"size:100;index"`
	UpdatedBy     string                    `json:"updatedBy" gorm:"size:100;index"`
}

func (*AnnouncementModel) TableName() string { return "announcements" }

func (m *AnnouncementModel) Normalize() {
	m.ScopeID = strings.TrimSpace(m.ScopeID)
	m.Title = strings.TrimSpace(m.Title)
	if m.ScopeType == "" {
		m.ScopeType = AnnouncementScopeWorld
	}
	if m.ContentFormat == "" {
		m.ContentFormat = AnnouncementContentRich
	}
	if m.Status == "" {
		m.Status = AnnouncementStatusDraft
	}
	if m.PopupMode == "" {
		m.PopupMode = AnnouncementPopupNone
	}
	if m.ReminderScope == "" {
		m.ReminderScope = AnnouncementReminderScopeLobbyOnly
	}
	if m.Version <= 0 {
		m.Version = 1
	}
}

type AnnouncementUserStateModel struct {
	StringPKBaseModel
	AnnouncementID  string     `json:"announcementId" gorm:"size:100;uniqueIndex:idx_ann_user_state,priority:1"`
	UserID          string     `json:"userId" gorm:"size:100;uniqueIndex:idx_ann_user_state,priority:2"`
	LastSeenVersion int        `json:"lastSeenVersion" gorm:"default:0"`
	LastPopupAt     *time.Time `json:"lastPopupAt"`
	AckVersion      int        `json:"ackVersion" gorm:"default:0"`
	AckAt           *time.Time `json:"ackAt"`
}

func (*AnnouncementUserStateModel) TableName() string { return "announcement_user_states" }
